File: crates/trusty-agents/ui/src/lib/costs.ts
// Costs tab API client + pure presentation helpers (#4098, COST-09).
//
// Why: `CostsView.svelte` must distinguish four backend states — costs,
// nothing recorded, nothing dispatched, and a fault — and it must never render
// a confident `$0.00` over a missing usage log. Keeping the fetch + the state
// discrimination here (rather than inline in the component) makes both
// unit-testable without mounting Svelte, and keeps the component to rendering.
//
// Note the deliberate deviation from the sibling clients (`workstreams.ts`,
// `agentConfig.ts`): those swallow failures and return `[]`, which is right for
// a convenience sidebar. It is WRONG here. An empty array is
// indistinguishable from "$0 spent", which is the single claim this tab must
// never make falsely, so `fetchCosts` returns a discriminated result and lets
// the view say which failure it hit.
//
// What: `CostRow`/`CostSummary` mirror `usage::aggregate`'s serialized shape;
// `fetchCosts` wraps `GET /api/costs`; `formatUsd`/`formatCount`/`barPercent`
// are pure formatters the view and its tests share.
// Test: `costs.test.ts`.

import { apiBase } from './api-config';
import { getCurrentApiToken } from '../stores/app';

/** One aggregate bucket. `key` is an agent name, model id, or `YYYY-MM-DD`. */
export interface CostRow {
  key: string;
  input_tokens: number;
  output_tokens: number;
  cost_usd: number;
  dispatch_count: number;
  duration_ms: number;
}

/** The `GET /api/costs` payload. Mirrors `usage::aggregate::CostSummary`. */
export interface CostSummary {
  available: boolean;
  source: string;
  records: number;
  malformed_lines: number;
  first_ts: string | null;
  last_ts: string | null;
  window_days: number | null;
  totals: CostRow;
  by_agent: CostRow[];
  by_model: CostRow[];
  by_date: CostRow[];
  /** Present only when `available` is false — why there is nothing to show. */
  reason?: string;
}

/**
 * The four states the view renders, kept as a discriminated union so no branch
 * can be silently skipped:
 *  - `ok`        — real aggregated costs (possibly zero rows, but a real log).
 *  - `no-data`   — no usage log exists yet. NOT `$0.00`.
 *  - `error`     — the log exists and could not be read, or the request failed.
 */
export type CostsState =
  | { kind: 'ok'; summary: CostSummary }
  | { kind: 'no-data'; reason: string; source: string }
  | { kind: 'error'; message: string };

function authHeaders(): Record<string, string> {
  const token = getCurrentApiToken();
  return token ? { Authorization: `Bearer ${token}` } : {};
}

/**
 * `GET /api/costs`, mapped to a [`CostsState`].
 *
 * Why: See the module note — this deliberately does NOT degrade a failure into
 * an empty dataset. A cost figure the operator cannot trust is worse than no
 * figure, so every non-success path keeps its identity all the way to the view.
 * What: Issues the request with the optional `?days=` window, maps a non-OK
 * status or a thrown network error to `error`, and an `available: false` body
 * to `no-data` carrying the server's own reason string.
 * Test: `fetchCosts` cases in `costs.test.ts`.
 */
export async function fetchCosts(days?: number): Promise<CostsState> {
  const query = days && days > 0 ? `?days=${days}` : '';
  try {
    const r = await fetch(`${apiBase()}/api/costs${query}`, { headers: authHeaders() });
    const body = (await r.json().catch(() => null)) as (CostSummary & { error?: string }) | null;
    if (!r.ok) {
      return { kind: 'error', message: body?.error ?? `GET /api/costs failed: ${r.status}` };
    }
    if (!body) return { kind: 'error', message: 'GET /api/costs returned an unreadable body' };
    if (body.available === false) {
      return {
        kind: 'no-data',
        reason: body.reason ?? 'no usage has been recorded for this project yet',
        source: body.source ?? '',
      };
    }
    return { kind: 'ok', summary: body };
  } catch (e) {
    return { kind: 'error', message: e instanceof Error ? e.message : String(e) };
  }
}

/**
 * Format a USD amount at a precision that does not imply false certainty.
 *
 * Why: Sub-cent dispatch costs round to `$0.00` at two decimals, which reads as
 * "free" for a column of real spend. The statusline already solved this the
 * same way (`repl::tui::status::format_cost_value`), so the two surfaces agree.
 * What: 4 decimals below $0.01, 2 decimals at or above.
 * Test: `formatUsd` cases in `costs.test.ts`.
 */
export function formatUsd(cost: number): string {
  if (!Number.isFinite(cost)) return '—';
  return cost < 0.01 ? `$${cost.toFixed(4)}` : `$${cost.toFixed(2)}`;
}

/** Compact integer with a `k`/`M` suffix — token counts get long fast. */
export function formatCount(n: number): string {
  if (!Number.isFinite(n)) return '—';
  if (n >= 1_000_000) return `${(n / 1_000_000).toFixed(1)}M`;
  if (n >= 1_000) return `${(n / 1_000).toFixed(1)}k`;
  return `${n}`;
}

/**
 * Bar width as a percentage of the largest row in the same breakdown.
 *
 * Why: The bars are relative magnitude, not absolute scale, so the chart stays
 * readable whether the day's spend is cents or hundreds of dollars.
 * What: `cost / max * 100`, clamped to [0, 100]; `0` when `max` is not a
 * positive finite number (an all-zero breakdown draws no bars rather than
 * dividing by zero into `NaN%`).
 * Test: `barPercent` cases in `costs.test.ts`.
 */
export function barPercent(cost: number, max: number): number {
  if (!Number.isFinite(cost) || !Number.isFinite(max) || max <= 0) return 0;
  return Math.max(0, Math.min(100, (cost / max) * 100));
}

File: crates/trusty-agents/src/api/server/costs.rs
//! `GET /api/costs` — aggregated usage cost for the Costs tab (#4098).
//!
//! Why: The Costs tab (COST-09, formerly #4108) needs totals plus a breakdown
//! by agent, model and day, and nothing exposed the per-dispatch usage log over
//! HTTP. This route is that surface. It owns no arithmetic of its own: the fold
//! is `usage::aggregate::aggregate_usage` and the rates are
//! `perf::pricing::cost_usd`, so the API can never report a number the CLI or a
//! future report would disagree with.
//!
//! **Honesty rules, and why each is load-bearing.** A cost view is only worth
//! having if a number on it is a claim you can act on, so every state that
//! ISN'T "here are your costs" gets its own distinguishable answer:
//!
//! - **No log yet → `200` with `available: false` and a `reason`, never a zero
//!   summary.** `$0.00` and "nothing has been recorded" are different claims,
//!   and the first one, made wrongly, is the failure mode that makes an
//!   operator stop trusting the whole tab. The GUI renders the reason.
//! - **A log that exists but cannot be read → `500`.** That is a real fault
//!   (permissions, a truncated mount) and is not degraded into "no data".
//! - **Unparseable lines → `200`, aggregated without them, with
//!   `malformed_lines > 0`.** Partial data is still useful; presenting a short
//!   total as a whole one is not. The GUI warns when the count is non-zero.
//! - **An empty-but-present log → `200`, `available: true`, `records: 0`.** The
//!   project has state and simply has not dispatched — a real, different answer
//!   from both of the above.
//!
//! Response shape (a single object, not COST-08's bare array):
//!
//! ```json
//! {
//!   "available": true,
//!   "source": "/proj/.trusty-agents/state/usage.jsonl",
//!   "records": 128, "malformed_lines": 0,
//!   "first_ts": "...", "last_ts": "...", "window_days": 7,
//!   "totals":   { "key": "total", "input_tokens": 0, "output_tokens": 0,
//!                 "cost_usd": 0.0, "dispatch_count": 0, "duration_ms": 0 },
//!   "by_agent": [ ... ], "by_model": [ ... ], "by_date": [ ... ]
//! }
//! ```
//!
//! Two deliberate deviations from COST-08's stated criteria, both consequences
//! of building on read-time aggregation rather than #4104's unimplemented
//! rollup tables (see `usage::aggregate`'s module doc and the PR body):
//!
//! - `?range=day|week|month` is not implemented — week/month rollups are
//!   COST-07 (#4105) and there is nothing to query. `?days=<n>` narrows the
//!   window instead, and `by_date` gives daily granularity.
//! - `?group_by=` is not implemented. All three breakdowns ship in one payload
//!   so the tab switches grouping without a refetch AND so the three can never
//!   disagree — they are folds of one pass over one set of rows.
//!
//! What: [`get_costs`] is the axum handler; [`costs_at`] is the testable core
//! that takes an explicit project dir.
//! Test: `super::tests::costs`.

use axum::{
    Json,
    extract::Query,
    http::StatusCode,
    response::{IntoResponse, Response},
};
use serde::Deserialize;
use serde_json::json;

use crate::usage::aggregate::{AggregateError, aggregate_usage};

/// Upper bound on `?days=`, so a hostile or fat-fingered value cannot ask for
/// a window wider than any plausible log. Purely defensive — the fold reads the
/// whole file regardless; this only bounds the reported `window_days`.
const MAX_WINDOW_DAYS: u32 = 3650;

/// Query string for `GET /api/costs`.
///
/// Why: One optional knob. See the module doc for why `range`/`group_by` from
/// COST-08 are absent rather than stubbed — a parameter that silently ignores
/// its value is worse than one that isn't there.
/// What: `days` — keep only rows within N days of the newest recorded row.
/// Omitted (or 0) means "everything".
/// Test: `costs_window_narrows_the_report`.
#[derive(Debug, Default, Deserialize)]
pub(super) struct CostsQuery {
    /// Window in days, anchored on the newest recorded row.
    #[serde(default)]
    days: Option<u32>,
}

/// `GET /api/costs` — HTTP entry point.
///
/// Why/What: see the module doc. The project root is the daemon's CWD, matching
/// `usage::project_dir`'s convention and the sibling routes' resolution.
/// Test: `super::tests::costs::costs_route_is_wired_into_router`.
pub(super) async fn get_costs(Query(q): Query<CostsQuery>) -> Response {
    costs_at(&crate::usage::project_dir(), q.days)
}

/// Aggregate `project_dir`'s usage log into an HTTP response.
///
/// Why: Split from the axum shim so the four states this route distinguishes
/// (recorded / not recorded / unreadable / malformed) are testable against a
/// tempdir without standing up a server or mutating the process CWD.
/// What: Calls `aggregate_usage`, maps `NotRecorded` to a `200` "no data"
/// envelope and `Read` to a `500`, and otherwise serializes the summary with
/// `available: true` spliced in.
/// Test: `costs_reports_no_data_for_missing_log`,
/// `costs_returns_totals_and_breakdowns`, `costs_surfaces_malformed_line_count`,
/// `costs_reports_an_unreadable_log_as_a_server_error`.
pub(super) fn costs_at(project_dir: &std::path::Path, days: Option<u32>) -> Response {
    let window = days.filter(|d| *d > 0).map(|d| d.min(MAX_WINDOW_DAYS));
    match aggregate_usage(project_dir, window) {
        Ok(summary) => {
            let mut body = match serde_json::to_value(&summary) {
                Ok(serde_json::Value::Object(map)) => map,
                _ => {
                    // Unreachable: `CostSummary` is a plain struct. Answering
                    // 500 rather than unwrapping keeps the daemon alive if it
                    // ever becomes reachable.
                    return (
                        StatusCode::INTERNAL_SERVER_ERROR,
                        Json(json!({ "error": "could not serialize cost summary" })),
                    )
                        .into_response();
                }
            };
            body.insert("available".to_string(), json!(true));
            (StatusCode::OK, Json(serde_json::Value::Object(body))).into_response()
        }
        // Not an error: a project that has never dispatched has no log. The
        // caller gets an explicit "nothing recorded", never a zero total.
        Err(AggregateError::NotRecorded { path }) => (
            StatusCode::OK,
            Json(json!({
                "available": false,
                "reason": "no usage has been recorded for this project yet",
                "source": path.display().to_string(),
                "records": 0,
                "malformed_lines": 0,
                "window_days": window,
                "totals": crate::usage::aggregate::CostRow {
                    key: "total".to_string(),
                    ..Default::default()
                },
                "by_agent": [], "by_model": [], "by_date": [],
            })),
        )
            .into_response(),
        // A log that exists but will not read is a real fault. Degrading it to
        // "no data" would hide a broken cost trail behind an innocent message.
        Err(e @ AggregateError::Read { .. }) => {
            tracing::warn!(error = %e, "costs: usage log unreadable");
            (
                StatusCode::INTERNAL_SERVER_ERROR,
                Json(json!({ "available": false, "error": e.to_string() })),
            )
                .into_response()
        }
    }
}
